File: src/main/java/xyz/cofe/jtfm/gr/Point.java
package xyz.cofe.jtfm.gr;

import com.googlecode.lanterna.TerminalPosition;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Objects;
import java.util.Optional;

/**
 * Точка
 */
public final class Point {
    /** Координата */
    public final int x;

    /** Координата */
    public final int y;

    /**
     * Конструктор
     * @param x координата
     * @param y координата
     */
    public Point( int x, int y ){
        this.x = x;
        this.y = y;
    }

    /**
     * Конструктор x=0, y=0
     */
    public Point(){
        this.x = 0;
        this.y = 0;
    }

    public Point( @NonNull TerminalPosition p ){
        this.x = p.getColumn();
        this.y = p.getRow();
    }

    public TerminalPosition toTerminalPosition(){
        return new TerminalPosition(x,y);
    }

    /**
     * Смещение точки
     * @param p размер сдвига
     * @return смещенная точка (клон)
     */
    public Point translate( @NonNull Point p ){
        return new Point(x+p.x, y+p.y );
    }

    /**
     * Смещение точки
     * @param x_off смещение
     * @param y_off смещение
     * @return смещенная точка (клон)
     */
    public Point translate( int x_off, int y_off ){
        return new Point(x+x_off, y+y_off );
    }

    /**
     * Инверсия координат
     * @param x_inv инвертировать x
     * @param y_inv инвертировать y
     * @return Инвертированная точка
     */
    public Point invert(boolean x_inv, boolean y_inv){
        return new Point(x_inv ? -x : x, y_inv ? -y : y);
    }

    /**
     * Инверсия координат
     * @return Инвертированная точка
     */
    public Point invert() { return invert(true, true); }

    @Override
    public boolean equals( @Nullable Object o ){
        if( this == o ) return true;
        if( o == null || getClass() != o.getClass() ) return false;
        Point point = (Point) o;
        return x == point.x && y == point.y;
    }

    @Override
    public int hashCode(){
        return Objects.hash(x, y);
    }

    @Override
    public String toString(){
        return "("+x+";"+y+")";
    }

    public static Optional<Point> parse( @NonNull String txt ){
        //noinspection ConstantConditions
        if( txt==null )throw new IllegalArgumentException( "txt==null" );
        if( !txt.startsWith("(") )return Optional.empty();
        if( !txt.endsWith(")") )return Optional.empty();
        if( txt.length()<3 )return Optional.empty();
        var txt1 = txt.substring(1,txt.length()-1);

        var xy = txt1.split(";");
        if( xy.length<2 )return Optional.empty();

        try{
            int x = Integer.parseInt(xy[0]);
            int y = Integer.parseInt(xy[1]);
            return Optional.of(new Point(x,y));
        } catch( NumberFormatException e ){
            return Optional.empty();
        }
    }
}

File: jtfm-scala/src/main/scala/xyz/cofe/jtfm/gr/Point.scala
package xyz.cofe.jtfm.gr

import com.googlecode.lanterna.TerminalPosition
import com.googlecode.lanterna.input.MouseAction

/**
 * Точка в прострастве
 */
final case class Point(x:Int, y:Int) {
    /**
     * Перенос точки
     */
    def translate(p:Point):Point = Point(x=x+p.x, y=y+p.y)

    /**
     * Перенос точки
     */
    def translate(off_x:Int, off_y:Int):Point = Point(off_x+x, off_y+y)

    /**
     * Инверсия точки
     */
    def invert(x_inv:Boolean, y_inv:Boolean):Point = Point(if(x_inv)-x else x, if(y_inv)-y else y)
    
    /**
     * Разница
     * @param p точка
     * @return разница
     */
    def diff( p:Point ):Point = Point( x-p.x, y-p.y )

    /**
     * Инверсия точки
     */
    def invert():Point = invert(true,true)

    /**
     * Возвращает текстовое представление
     * @return для x=5, y=3 вернет "(5;3)"
     */
    override def toString():String = s"($x;$y)"
}

object Point {
    def apply(str:String):Either[Point,String] = parse(str).left.map( _._1 )
    def apply(pos:TerminalPosition):Point = Point(pos.getColumn, pos.getRow)
    def apply(pos:(Int,Int)):Point = Point(pos._1, pos._2)
    implicit class TermPosOps(val term_pos:TerminalPosition) {
        def toPoint():Point = Point(term_pos.getColumn, term_pos.getRow)
    }
    implicit class PointTermPosOps(val p:Point) {
        def toTerminalPosition():TerminalPosition = new TerminalPosition(p.x, p.y)
    }
    private lazy val parse_pattern = java.util.regex.Pattern.compile("\\((?<x>\\d+);(?<y>\\d+)\\)")

    def parse(str:String, from:Int=0):Either[(Point,Int),String] = {
        if(str==null){
            Right("argument is null")
        }else if( from>=str.length || from<0 ){
            Right("argument from out of range")
        }else{
            val rest = str.substring(from)
            val m = parse_pattern.matcher(rest)
            if( m.matches ){
                val x = m.group("x").toInt
                val y = m.group("y").toInt
                val next_off = m.end+from
                Left((Point(x,y),next_off))
            }else{
                Right(s"'$str' not match (\\d+;\\d+)")
            }
        }
    }

    implicit def fromPointToTerminalPosition( p:Point ):TerminalPosition =
        new TerminalPosition(p.x, p.y)

    implicit def terminalPosition2Point( p:TerminalPosition ):Point =
        Point(p.getColumn, p.getRow)

    implicit def mouseAction2Point( ma:MouseAction ):Point =
        Point(ma.getPosition.getColumn, ma.getPosition.getRow)

    }
